### PR TITLE
Improvement controller session

### PR DESCRIPTION
**Pull request by morrieinmaas**
## Description
Refactor of Agent controller to be more concise, have doc strings and handle session and token and api-key in header in a better way.

## How has this been tested?
- In notebooks

### CODE REVIEW

File: libs/aries-basic-controller/aries_basic_controller/aries_controller.py [hunk 1]
```diff
@@ -25,51 +26,84 @@
...
+        Initialise the mediation interface (default is False)
+    issuer : bool
+        Initialise the issuer interface (defautl is True)
+    action_menu : bool
+        Initialise the action menu interface (default is True)
+    revocations : bool
+        Initialise revocation interface for credentials (default is True)
+    api_key : str 
+        The API key (default is None)
+    tenant_jwt: str
```
Comment (wip-abramson) [inline libs/aries-basic-controller/aries_basic_controller/aries_controller.py:62]: FYI its tennant ;)

File: libs/aries-basic-controller/aries_basic_controller/aries_controller.py [hunk 2]
```diff
@@ -25,51 +26,84 @@
...
+        The exposed port for webhooks on the host
+    admin_url : str
+        The URL for the Admin API
+    webhook_base : str
+        The base url for webhooks (default is "")
+    connections : bool
+        Specify whether to create connecitons (default is True)
+    messaging : bool
+        Initialise the messaging interface (default is True)
+    multitenant : bool
```
Comment (wip-abramson) [inline libs/aries-basic-controller/aries_basic_controller/aries_controller.py:50]: and again tennant

File: libs/aries-basic-controller/aries_basic_controller/aries_controller.py [hunk 3]
```diff
@@ -25,51 +26,84 @@
...
-        tennant_jwt: str = None,
-    ):
-
+    webhook_host: str
+    webhook_port: int
+    admin_url: str
+    webhook_base: str = ""
+    connections: bool = True
+    messaging: bool = True
+    multitenant: bool = False
```
Comment (wip-abramson) [inline libs/aries-basic-controller/aries_basic_controller/aries_controller.py:74]: We should really make this multitennant. But means updating the tutorials sorry

File: libs/aries-basic-controller/aries_basic_controller/aries_controller.py [hunk 4]
```diff
@@ -25,51 +26,84 @@
...
+    mediation: bool = False
+    issuer: bool = True
+    action_menu: bool = True
+    revocations: bool = True
+    api_key: str = None
+    tenant_jwt: str = None
```
Comment (wip-abramson) [inline libs/aries-basic-controller/aries_basic_controller/aries_controller.py:80]: Also this

File: libs/aries-basic-controller/aries_basic_controller/aries_controller.py [hunk 5]
```diff
@@ -141,8 +134,63 @@ def update_tennant_jwt(self, tennant_jwt):
                 self.client_session
             )
 
+    
+    def update_tenant_jwt(self, tenant_jwt: str):
```
Comment (wip-abramson) [inline libs/aries-basic-controller/aries_basic_controller/aries_controller.py:138]: tennant. Seems like we just need a find and replace?

File: libs/aries-basic-controller/aries_basic_controller/aries_controller.py [hunk 6]
```diff
@@ -154,40 +202,91 @@ def register_listeners(self, listeners, defaults=True):
...
+            A response with status 200
+        """
         topic = request.match_info["topic"]
         payload = await request.json()
         await self._handle_webhook(topic, payload)
         return web.Response(status=200)
 
+
     async def _handle_webhook(self, topic, payload):
+        """Helper handling a webhook
```
Comment (wip-abramson) [inline libs/aries-basic-controller/aries_basic_controller/aries_controller.py:274]: I would say. Handles webhook by publishing payload to any functions subscribing to topic through PubSub. Or something like that

File: tutorials/4. Multitenancy/notebooks/basewallet/Alice/Part 2 - Mediation of communication - Alice.ipynb [hunk 7]
```diff
@@ -119,7 +119,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "agent_controller.update_tennant_jwt(alice_jwt)"
+    "agent_controller.update_tenant_jwt(alice_jwt)"
```
Comment (wip-abramson) [inline tutorials/4. Multitenancy/notebooks/basewallet/Alice/Part 2 - Mediation of communication - Alice.ipynb:122]: Lol you changed it back?

File: tutorials/4. Multitenancy/notebooks/basewallet/Part 1 - Subwallet Administration.ipynb [hunk 8]
```diff
@@ -7,21 +7,21 @@
    "source": [
     "# Part 1 - Exploring Sub Wallet Management\n",
     "\n",
-    "This agent has been initialised using the multitennant flag. This means a single ACA-Py instance can be used to manage multiple subwallets. Each tennant gets their own encrypted data storage for managing their own connections, credentials and interactions etc. \n",
+    "This agent has been initialised using the multitenant flag. This means a single ACA-Py instance can be used to manage multiple subwallets. Each tenant gets their own encrypted data storage for managing their own connections, credentials and interactions etc. \n",
```
Comment (wip-abramson) [inline tutorials/4. Multitenancy/notebooks/basewallet/Part 1 - Subwallet Administration.ipynb:10]: Im confused. Pretty sure its multitennant, but now thinking maybe I am wrong

Comment (wip-abramson) [reply tutorials/4. Multitenancy/notebooks/basewallet/Part 1 - Subwallet Administration.ipynb:10]: Okay turns out I was wrong you were right hehe

**Review by wip-abramson — APPROVED** [review]

File: libs/aries-basic-controller/aries_basic_controller/aries_controller.py (hunk 6 above)
Comment (wip-abramson) [reply libs/aries-basic-controller/aries_basic_controller/aries_controller.py:274]: Actually I approved but I think this still would be a good change to make @morrieinmaas

**Comment by morrieinmaas** [comment]
I was confused also. thought it's double n, but according to the dictionary it's tenant. Silly me

**Pull request merged**